crates/cli/src/native_query/type_constraint.rs

```diff
@@ -16,28 +16,46 @@ impl TypeVariable {
     pub fn new(id: u32, variance: Variance) -> Self {
         TypeVariable { id, variance }
     }
+
+    pub fn is_covariant(self) -> bool {
+        matches!(self.variance, Variance::Covariant)
+    }
+
+    pub fn is_contravariant(self) -> bool {
+        matches!(self.variance, Variance::Contravariant)
+    }
+}
+
+impl std::fmt::Display for TypeVariable {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "${}", self.id)
+    }
 }
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum Variance {
     Covariant,
     Contravariant,
+    Invariant,
 }
 
 /// A TypeConstraint is almost identical to a [configuration::schema::Type], except that
 /// a TypeConstraint may reference type variables.
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub enum TypeConstraint {
     // Normal type stuff - except that composite types might include variables in their structure.
     ExtendedJSON,
     Scalar(BsonScalarType),
     Object(ObjectTypeName),
     ArrayOf(Box<TypeConstraint>),
-    Nullable(Box<TypeConstraint>),
     Predicate {
         object_type_name: ObjectTypeName,
     },
 
+    // Complex types
+    Union(BTreeSet<TypeConstraint>),
+    OneOf(BTreeSet<TypeConstraint>), // unlike union type should be only one of these, but we don't know which one
```

Comment:
Could you provide a concrete example here in a comment perhaps? This may be entirely a reflection of my lack of understanding around Mongo's semantics, but I can't get my head around when this would be used.

Reply:
That's a good idea, I'll add more details in a comment.

OneOf is helpful mostly for numeric inputs, where we're expecting a numeric type, but we don't know specifically which one. We probably want to infer a specific numeric type instead of a union of all numeric types, which would present as `ExtendedJSON` (the top type) in the API which isn't as helpful.

OTOH my thinking is Union is for cases where we have specific evidence a type variable is associated with multiple distinct concrete types. I'm mainly using Union to indicate nullability.

Another way to handle OneOf that could be less handwavy could be to use `SubtypeOf(Union(...))`. But I'm not sure exactly how the semantics of that would play out.